interpreter/attributes.go

```diff
@@ -0,0 +1,844 @@
+func (r *attrFactory) ConditionalAttribute(id int64, expr Interpretable, t, f Attribute) Attribute {
+	return &conditionalAttribute{
+		id:      id,
+		expr:    expr,
+		truthy:  t,
+		falsy:   f,
+		adapter: r.adapter,
+		res:     r,
+	}
+}
+
+// OneofAttribute collects variants of unchecked AbsoluteAttribute values which could either be
+// direct variable accesses or some combination of variable access with qualification.
+func (r *attrFactory) OneofAttribute(id int64, name string) Attribute {
+	return &oneofAttribute{
+		id: id,
+		attrs: []*absoluteAttribute{
+			&absoluteAttribute{
+				id:             id,
+				namespaceNames: r.pkg.ResolveCandidateNames(name),
+				qualifiers:     []Qualifier{},
+				provider:       r.provider,
+				adapter:        r.adapter,
+				res:            r,
+			},
+		},
+		adapter:  r.adapter,
+		provider: r.provider,
+		res:      r,
+	}
+}
+
+// RelativeAttribute refers to an expression and an optional qualifier path.
+func (r *attrFactory) RelativeAttribute(id int64, operand Interpretable) Attribute {
+	return &relativeAttribute{
+		id:         id,
+		operand:    operand,
+		qualifiers: []Qualifier{},
+		adapter:    r.adapter,
+		res:        r,
+	}
+}
+
+// NewQualifier is an implementation of the AttributeFactory interface.
+func (r *attrFactory) NewQualifier(objType *exprpb.Type,
+	qualID int64,
+	val interface{}) (Qualifier, error) {
+	// Before creating a new qualifier check to see if this is a protobuf message field access.
+	// If so, use the precomputed GetFrom qualification method ratehr than the standard
+	// stringQualifier.
+	str, isStr := val.(string)
+	if isStr && objType != nil && objType.GetMessageType() != "" {
+		ft, found := r.provider.FindFieldType(objType.GetMessageType(), str)
+		if found && ft.IsSet != nil && ft.GetFrom != nil {
+			return &fieldQualifier{
+				id:        qualID,
+				Name:      str,
+				FieldType: ft,
+				adapter:   r.adapter,
+			}, nil
+		}
+	}
+	return newQualifier(r.adapter, qualID, val)
+}
+
+type absoluteAttribute struct {
+	id             int64
+	namespaceNames []string
+	qualifiers     []Qualifier
+	adapter        ref.TypeAdapter
+	provider       ref.TypeProvider
+	res            AttributeFactory
+}
+
+// ID implements the Attribute interface method.
+func (a *absoluteAttribute) ID() int64 {
+	return a.id
+}
+
+// AddQualifier implements the Attribute interface method.
+func (a *absoluteAttribute) AddQualifier(qual Qualifier) (Attribute, error) {
+	a.qualifiers = append(a.qualifiers, qual)
+	return a, nil
+}
+
+// Resolve iterates through the namespaced variable names until one is found in the Activation,
+// and the the standard qualifier resolution logic is applied.
+//
+// If the variable name is not found an error is returned.
+func (a *absoluteAttribute) Resolve(vars Activation) (interface{}, error) {
+	for _, nm := range a.namespaceNames {
+		op, found := vars.ResolveName(nm)
+		_, isUnk := op.(types.Unknown)
+		if found && !isUnk {
```

Comment:
You're precisely right that the variable may be getting shadowed by a type. This happens when using the `UnknownActivation` which will return that every name is found and unknown, even when the identifier could actually be a well-established type value. I've kept the logic, but documented why it is this way.

Reply:
Then consider either:

- drop UnknownActivation and PartialActivation right now, since they don't seem to be widely used, or
- check to see whether vars is a partialActivation and handle as a temporary special case within the ideal control-flow, rather than risking spread of the bad precedence for the general case.

Reply:
The `UnknownActivation` represents the pre-PR contract for `Activation`. I happen to know of a use case that depends on the ability to do this and would like to ensure that a sync of the latest code can still work correctly with it. 

The `Activation` could be implemented by anyone (and I have leveraged the interface heavily elsewhere). Being able to return unknown for a given variable name is an important capability and I don't particularly want to special case the handling of the response based on the concrete, private, implementation.